File: concurrency/01114_print_in_order/sem.c
#define STATES_CNT 3

typedef struct {
    // User defined data may be declared here.
    sem_t states[STATES_CNT];
} Foo;

// Function Declaration, do not remove
void printFirst();
void printSecond();
void printThird();

Foo* fooCreate() {
    Foo* obj = (Foo*) malloc(sizeof(Foo));
    // Initialize user defined data here.
    sem_init(&obj->states[0], 0, 1);
    for (int i = 1; i < STATES_CNT; i += 1) {
        sem_init(&obj->states[i], 0, 0);
    }
    return obj;
}

void first(Foo* obj) {
    sem_wait(&obj->states[0]);
    // printFirst() outputs "first". Do not change or remove this line.
    printFirst();
    sem_post(&obj->states[1]);
}

void second(Foo* obj) {
    sem_wait(&obj->states[1]);
    // printSecond() outputs "second". Do not change or remove this line.
    printSecond();
    sem_post(&obj->states[2]);
}

void third(Foo* obj) {
    sem_wait(&obj->states[2]);
    // printThird() outputs "third". Do not change or remove this line.
    printThird();
    sem_post(&obj->states[0]);
}

void fooFree(Foo* obj) {
    // User defined data may be cleaned up here.
    free(obj);
}

File: concurrency/01114_print_in_order/sem.rs
#[derive(Clone)]
struct Sem {
    count: Arc<(Mutex<usize>, Condvar)>,
}

impl Sem {
    fn new(val: usize) -> Self {
        Self {
            count: Arc::new((Mutex::new(val), Condvar::new())),
        }
    }

    fn wait(&self) {
        let (lock, cv) = &*self.count;
        let mut count = lock.lock().unwrap();
        while *count == 0 {
            count = cv.wait(count).unwrap();
        }
        *count = 0;
    }

    fn post(&self) {
        let (lock, cv) = &*self.count;
        let mut count = lock.lock().unwrap();
        *count = 1;
        cv.notify_one();
    }
}


struct Foo {
    states: [Sem; 3],
}

impl Foo {
    fn new() -> Self {
        Foo {
            states: [Sem::new(1), Sem::new(0), Sem::new(0)],
        }
    }

    fn first<F>(&self, print_first: F)
    where
        F: FnOnce(),
    {
        self.states[0].wait();
        // Do not change this line
        print_first();
        self.states[1].post();
    }

    fn second<F>(&self, print_second: F)
    where
        F: FnOnce(),
    {
        self.states[1].wait();
        // Do not change this line
        print_second();
        self.states[2].post();
    }

    fn third<F>(&self, print_third: F)
    where
        F: FnOnce(),
    {
        self.states[2].wait();
        // Do not change this line
        print_third();
        self.states[0].post();
    }
}
